adapters/repos/db/vector/cuvs/index.go

```diff
@@ -0,0 +1,739 @@
+		targetVector:   cfg.TargetVector,
+		logger:         logger,
+		distanceMetric: cuvs.DistanceL2, // TODO: make configurable
+		store:          store,
+		rootPath:       cfg.RootPath,
+
+		idCuvsIdMap:    make(map[uint32]uint64),
+		cuvsPoolMemory: cfg.CuvsPoolMemory,
+	}
+
+	println("pre init buckets")
+
+	if err := index.initBuckets(context.Background()); err != nil {
+		return nil, fmt.Errorf("init cuvs index buckets: %w", err)
+	}
+	println("post init buckets")
+
+	if err := index.initMetadata(); err != nil {
+		return nil, err
+	}
+
+	return index, nil
+}
+
+func byteSliceFromFloat32Slice(vector []float32, slice []byte) []byte {
+	for i := range vector {
+		binary.LittleEndian.PutUint32(slice[i*4:], math.Float32bits(vector[i]))
+	}
+	return slice
+}
+
+func shouldExtend(index *cuvs_index, num_new uint64) bool {
+	if num_new > 100 {
+		// println("num_new is greater than 100; rebuilding")
+		return false
+	}
+
+	if index.count < 300_000 {
+		// println("index count is less than 300_000; rebuilding")
+		return false
+	}
+
+	if index.dlpackTensor == nil {
+		// println("dlpack tensor is nil; rebuilding")
+		return false
+	}
+
+	percentNewVectors := (float32(num_new+index.cuvsExtendCount) / float32(index.count+num_new)) * 100
+
+	// why 20? https://weaviate-org.slack.com/archives/C05V3MGDGQY/p1722897390825229?thread_ts=1722894509.398509&cid=C05V3MGDGQY
+	if percentNewVectors > 20 {
+		println("percentNewVectors is greater than 20; rebuilding")
+		return false
+	}
+
+	return true
+}
+
+func Normalize_Temp(vector []float32) []float32 {
+	newVector := make([]float32, len(vector))
+	for i := range vector {
+		newVector[i] = vector[i] * 0.146
+	}
+	return newVector
+}
+
+func (index *cuvs_index) Add(ctx context.Context, id uint64, vector []float32) error {
+	return index.AddBatch(ctx, []uint64{id}, [][]float32{vector})
+}
+
+func (index *cuvs_index) AddBatch(ctx context.Context, id []uint64, vectors [][]float32) error {
+	index.Lock()
+	defer index.Unlock()
+
+	if len(id) != len(vectors) {
+		return errors.Errorf("ids and vectors sizes does not match")
+	}
+	if len(id) == 0 {
+		return errors.Errorf("insertBatch called with empty lists")
+	}
+
+	index.trackDimensionsOnce.Do(func() {
+		size := int32(len(vectors[0]))
+		atomic.StoreInt32(&index.dims, size)
+		err := index.setDimensions(size)
+		if err != nil {
+			index.logger.WithError(err).Error("could not set dimensions")
+		}
+	})
+	for i := range vectors {
+		if len(vectors[i]) != int(index.dims) {
+			return errors.Errorf("insert called with a vector of the wrong size")
+		}
+	}
+
+	if index == nil {
+		return errors.New("cuvs index is nil")
+	}
+
+	for v := range vectors {
+		vectors[v] = Normalize_Temp(vectors[v])
+	}
+
+	// store in bucket
+	for i := range id {
+		slice := make([]byte, len(vectors[i])*4)
+		idBytes := make([]byte, 8)
+		binary.BigEndian.PutUint64(idBytes, id[i])
+		index.store.Bucket(index.getBucketName()).Put(idBytes, byteSliceFromFloat32Slice(vectors[i], slice))
```

Comment:
I wonder if we should more closely mirror the Dynamic index here where generally a flat index is used and then the cuVS index is rebuilt as needed based of the flat index.

Reply:
That's definitely a good idea - nvidia is planning to create something similar in the library itself, I'll see how far along this feature is